### DISCUSSION TITLE
Using openBLAS

### DISCUSSION THREAD

**Discussion by RajalakshmiSR** [microsoft/onnxruntime#6829] [Other Q&A]
There is a statement on the announcement page “In the next release, MKL-ML, openblas, and jemallac build options will be removed”.
Does this mean OpenBLAS GEMM functions will not be used in ONNX runtime? Or is it just the options in the build script that will be removed and ONNX Runtime can use OpenBLAS for SGEMM/DGEMM/CGEMM/ZGEMM?

**Comment by snnn** [comment]
>Does this mean OpenBLAS GEMM functions will not be used in ONNX runtime?

Right. By default we use our own GEMM implementation.  If a user need something different, he/she is encouraged to try the other execution providers(like DNNL) or build his own one. The default CPU execution providers only uses [mlas](https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/core/mlas) and eigen as the math libs.

**Reply by RajalakshmiSR** [reply]
Thanks for the reply. So If I have to use OpenBLAS, Is there any build variable that I can use to specify?

**Reply by snnn** [reply]
No. We have completely removed the related code. Based on the metrics we have, we believe our mlas performs better than OpenBLAS. If it is not the case,  we are happy to know.

**Reply by RajalakshmiSR** [reply]
For POWER architecture, there are optimized float/double/bfloat16 GEMM kernels in OpenBLAS that can give better results than default mlas. While adding optimized kernel in mlas for such architecture is one way of solving, allowing to use OpenBLAS would also help to handle it.

**Reply by snnn** [reply]
I believe @tracysh  has done some work for PowerPC in mlas.  Did you compare the perf? Is OpenBLAS faster?

**Reply by RajalakshmiSR** [reply]
I recently added POWER10(MMA) optimized GEMM kernels in OpenBLAS. That will be faster for POWER10 processors.

**Reply by RajalakshmiSR** [reply]
>  The default CPU execution providers only uses [mlas](https://github.com/microsoft/onnxruntime/tree/master/onnxruntime/core/mlas) and eigen as the math libs.

@snnn How does onnx runtime decide to use Eigen or MLAS path for matrix multiplication specifically GEMM? (assuming we have GEMM implementation in both eigen and MLAS)

**Reply by snnn** [reply]
It prefers MLAS when MLAS has a GEMM implementation on that platform.

**Reply by RajalakshmiSR** [reply]
@snnn Is there any switch like USE_EIGEN_FOR_BLAS where I can use Eigen for sgemm instead of MLAS for some testing?

**Reply by snnn** [reply]
We don't. But welcome to contribute.

**Reply by miantiao5115** [reply]
> @snnn Is there any switch like USE_EIGEN_FOR_BLAS where I can use Eigen for sgemm instead of MLAS for some testing?

Hi, have you tried to replace mlas with openblas successfully?
